planner/core/casetest/testdata/predicate_simplification_out.json

```diff
@@ -0,0 +1,183 @@
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "select f from t use index() where f not in (1,2,3) and f in (1,2,3) -- intersection of in and not in. Not done yet",
+        "Plan": [
+          "TableReader 0.00 root  data:Selection",
+          "└─Selection 0.00 cop[tikv]  in(test.t.f, 1, 2, 3), not(in(test.t.f, 1, 2, 3))",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      }
+    ]
+  },
+  {
+    "Name": "TestInListAndNotEqualSimplification",
+    "Cases": [
+      {
+        "SQL": "select f from t use index() where f <> 1 and f in (1,2,3) -- Simple case",
+        "Plan": [
+          "TableReader 20.00 root  data:Selection",
+          "└─Selection 20.00 cop[tikv]  in(test.t.f, 2, 3)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "select f from t use index() where f in (1,2,3) and f <> 2 -- Simple case with different order. Test case for issue 39676",
+        "Plan": [
+          "TableReader 20.00 root  data:Selection",
+          "└─Selection 20.00 cop[tikv]  in(test.t.f, 1, 3)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "select f from t use index() where f <> 3 and f in (1,2,3) and f <> 2 -- Multiple <> values. All are in inlist",
+        "Plan": [
+          "TableReader 10.00 root  data:Selection",
+          "└─Selection 10.00 cop[tikv]  in(test.t.f, 1)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "select f from t use index() where f in (1,2,3) and f <> 2 and f <> 99 -- Multiple <> values. Some are in inlist",
+        "Plan": [
+          "TableReader 20.00 root  data:Selection",
+          "└─Selection 20.00 cop[tikv]  in(test.t.f, 1, 3)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "select f from t use index() where f in (1,2,3) and f <> 5 and f <> 33  -- Multiple <> values. None are in inlist",
+        "Plan": [
+          "TableReader 30.00 root  data:Selection",
+          "└─Selection 30.00 cop[tikv]  in(test.t.f, 1, 2, 3)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "select f from t use index() where f <> 3 and f in (1,2,3) and f <> 1 and f <> 2 -- Multiple <> values and cover whole inlist. We keep at least one in inlist",
+        "Plan": [
+          "TableReader 10.00 root  data:Selection",
+          "└─Selection 10.00 cop[tikv]  in(test.t.f, 2)",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
```

Comment:
We get the wrong result in this case? If we keep `f in (2)`, we should also keep `f != 2`.

Reply:
You are right and we need to keep both. Good catch. Fixed.